bundles/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/HomekitAccessoryFactory.java

```diff
@@ -12,156 +12,417 @@
+     */
+    private static void addMandatoryCharacteristics(HomekitTaggedItem mainItem, List<HomekitTaggedItem> characteristics,
+            Item item, MetadataRegistry metadataRegistry) {
+        HomekitCharacteristicType[] requiredChar = mandatoryCharacteristics.get(mainItem.getAccessoryType());
+        for (Entry<HomekitAccessoryType, HomekitCharacteristicType> accessory : getAccessoryTypes(item,
+                metadataRegistry)) {
+            if (isRootAccessory(accessory) && (requiredChar != null)) {
+                Arrays.stream(requiredChar)
+                        .forEach(c -> characteristics.add(new HomekitTaggedItem(item, accessory.getKey(), c,
+                                mainItem.isGroup() ? (GroupItem) mainItem.getItem() : null,
+                                HomekitAccessoryFactory.getItemConfiguration(item, metadataRegistry))));
+            } else {
+                if (isMandatoryCharacteristic(mainItem.getAccessoryType(), legacyCheck(accessory.getValue())))
+                    characteristics
+                            .add(new HomekitTaggedItem(item, accessory.getKey(), legacyCheck(accessory.getValue()),
+                                    mainItem.isGroup() ? (GroupItem) mainItem.getItem() : null,
+                                    HomekitAccessoryFactory.getItemConfiguration(item, metadataRegistry)));
+            }
         }
     }
 
-    private static Map<HomekitCharacteristicType, Item> getCharacteristicItems(HomekitTaggedItem taggedItem) {
+    /**
+     * add optional characteristic for given accessory.
+     * 
+     * @param accessory accessory
+     * @param metadataRegistry metadata registry
+     */
+    private static void addOptionalCharacteristics(AbstractHomekitAccessoryImpl accessory,
+            MetadataRegistry metadataRegistry) {
+        Map<HomekitCharacteristicType, GenericItem> characteristics = getOptionalCharacteristics(
+                accessory.getRootAccessory(), metadataRegistry);
+        Service service = accessory.getPrimaryService();
+
+        characteristics.forEach((type, item) -> {
+            try {
+                logger.trace("adding optional characteristic: {} for item {}", type, item.getName());
+
+                final Characteristic characteristic = HomekitCharacteristicFactory.createCharacteristic(type, item,
+                        accessory.getUpdater());
+
+                service.getClass().getMethod("addOptionalCharacteristic", characteristic.getClass()). // find the
+                                                                                                      // corresponding
+                                                                                                      // add method at
+                                                                                                      // service and
+                                                                                                      // call it.
```

Comment:
to me it is a pretty elegant and future proof solution. 

alternative would a pretty big and ugly switch with 20 complex cases and growing. further, with reflection we will be able to add support for additional optional characteristics just by upgrading the underlaying lib without touching the addon code. 

important: this code is executed only at start of the binding and on changes in items configuration, so that it has no impact on runtime performance later.